File: R/R/parse_header.R
# Header parsing for Biobasis data files

source("utils.R")

parse_header <- function(file_path) {
  # Read first 4 lines
  lines <- readLines(file_path, n = 4)
  
  if (length(lines) < 4) {
    stop(paste("File", file_path, "has fewer than 4 header lines"))
  }
  
  # Line 2: Column names (split by comma, remove quotes)
  column_names <- trimws(gsub('"', '', strsplit(lines[2], ",")[[1]]))
  
  # Line 3: Units (split by comma, remove quotes)
  units_list <- trimws(gsub('"', '', strsplit(lines[3], ",")[[1]]))
  
  # Line 4: Statistics (split by comma, remove quotes)
  stats_list <- trimws(gsub('"', '', strsplit(lines[4], ",")[[1]]))
  
  # Create named vectors for units and stats
  units_dict <- setNames(rep("", length(column_names)), column_names)
  stats_dict <- setNames(rep("", length(column_names)), column_names)
  
  # Fill in available units and stats
  for (i in seq_along(column_names)) {
    if (i <= length(units_list)) {
      units_dict[column_names[i]] <- units_list[i]
    }
    if (i <= length(stats_list)) {
      stats_dict[column_names[i]] <- stats_list[i]
    }
  }
  
  log_message("DEBUG", paste("Parsed header from", file_path, ":", length(column_names), "columns"))
  
  return(list(
    columns = column_names,
    units = units_dict,
    stats = stats_dict
  ))
}

validate_header_consistency <- function(headers) {
  if (length(headers) == 0) {
    return()
  }
  
  reference_columns <- headers[[1]]$columns
  
  for (i in 2:length(headers)) {
    current_columns <- headers[[i]]$columns
    if (!identical(current_columns, reference_columns)) {
      log_message("WARNING", paste("Header", i, "has different columns than reference header"))
      
      ref_set <- reference_columns
      curr_set <- current_columns
      missing <- setdiff(ref_set, curr_set)
      extra <- setdiff(curr_set, ref_set)
      
      if (length(missing) > 0) {
        log_message("WARNING", paste("Missing columns:", paste(missing, collapse = ", ")))
      }
      if (length(extra) > 0) {
        log_message("WARNING", paste("Extra columns:", paste(extra, collapse = ", ")))
      }
    }
  }
}

get_timestamp_column_info <- function(column_names) {
  timestamp_candidates <- c('TIMESTAMP', 'timestamp', 'DateTime', 'datetime', 'TIME', 'time')
  
  for (candidate in timestamp_candidates) {
    if (candidate %in% column_names) {
      return(candidate)
    }
  }
  
  # If no standard timestamp column found, use first column
  if (length(column_names) > 0) {
    log_message("WARNING", paste("No standard timestamp column found, using first column:", column_names[1]))
    return(column_names[1])
  }
  
  stop("No columns found in header")
}

File: python/biobasis_merge_py/parse_header.py
"""Header parsing for Biobasis data files."""

import pandas as pd
from typing import Dict, List, Tuple
import logging

logger = logging.getLogger(__name__)


def parse_header(file_path: str) -> Tuple[List[str], Dict[str, str], Dict[str, str]]:
    """
    Parse 4-line header from Biobasis data file.
    
    Returns:
        Tuple of (column_names, units_dict, stats_dict)
    """
    try:
        with open(file_path, 'r') as f:
            lines = [f.readline().strip() for _ in range(4)]
        
        if len(lines) < 4:
            raise ValueError(f"File {file_path} has fewer than 4 header lines")
        
        # Line 2: Column names (split by comma)
        column_names = [col.strip('"') for col in lines[1].split(',')]
        
        # Line 3: Units (split by comma)
        units_list = [unit.strip('"') for unit in lines[2].split(',')]
        
        # Line 4: Statistics (split by comma)
        stats_list = [stat.strip('"') for stat in lines[3].split(',')]
        
        # Create dictionaries mapping column names to units/stats
        units_dict = {}
        stats_dict = {}
        
        for i, col_name in enumerate(column_names):
            if i < len(units_list):
                units_dict[col_name] = units_list[i]
            else:
                units_dict[col_name] = ""
                
            if i < len(stats_list):
                stats_dict[col_name] = stats_list[i]
            else:
                stats_dict[col_name] = ""
        
        logger.debug(f"Parsed header from {file_path}: {len(column_names)} columns")
        return column_names, units_dict, stats_dict
        
    except Exception as e:
        logger.error(f"Error parsing header from {file_path}: {e}")
        raise


def validate_header_consistency(headers: List[Tuple[List[str], Dict[str, str], Dict[str, str]]]) -> None:
    """Validate that all headers have consistent column structure."""
    if not headers:
        return
    
    reference_columns = headers[0][0]
    
    for i, (columns, _, _) in enumerate(headers[1:], 1):
        if columns != reference_columns:
            logger.warning(f"Header {i} has different columns than reference header")
            # Log the differences
            ref_set = set(reference_columns)
            curr_set = set(columns)
            missing = ref_set - curr_set
            extra = curr_set - ref_set
            if missing:
                logger.warning(f"Missing columns: {missing}")
            if extra:
                logger.warning(f"Extra columns: {extra}")


def get_timestamp_column_info(column_names: List[str]) -> str:
    """Find the timestamp column name."""
    timestamp_candidates = ['TIMESTAMP', 'timestamp', 'DateTime', 'datetime', 'TIME', 'time']
    
    for candidate in timestamp_candidates:
        if candidate in column_names:
            return candidate
    
    # If no standard timestamp column found, assume first column
    if column_names:
        logger.warning(f"No standard timestamp column found, using first column: {column_names[0]}")
        return column_names[0]
    
    raise ValueError("No columns found in header")
